lib/storage/src/upload/service-clients/Uploader.ts

```diff
@@ -0,0 +1,120 @@
+import {
+  AbortMultipartUploadRequest,
+  CompleteMultipartUploadRequest,
+  CreateMultipartUploadRequest,
+  ServiceOutputTypes,
+  UploadPartCommandInput,
+  UploadPartCommandOutput,
+} from "@aws-sdk/client-s3";
+import { EventEmitter } from "events";
+
+import { DataPart } from "../../data-chunk/yield-chunk";
+import { byteLength } from "../../utils/bytelength";
+import { UploadType } from "../Upload";
+
+export interface MultiPartIdentity {
+  UploadId: string;
+  Bucket: string;
+  Key: string;
+}
+
+export interface PartIdentity {
+  threadOrigin?: number;
+  PartNumber: number;
+  ETag: string;
+}
+
+export abstract class Uploader extends EventEmitter {
+  uploadEvent = "httpUploadProgress";
+  data: AsyncGenerator<DataPart, void, unknown>;
+  uploadedParts: PartIdentity[];
+  dataBytesUploaded: number;
+  configuration: UploadType.Configuration;
```

Comment:
`uploadEvent` can be a static constant. Also a lot of these members can be `private` or `readonly`. Same to other classes like `Upload`, `S3Uploader` and `S3ClientUploader`.